### PR TITLE
020502-draw-winners

### LINKED ISSUE #26: US 02.05.02
As an organizer I want to set the system to sample a specified number of attendees to register for the event.

### **Acceptance Criteria (For without DB)**
- [x] In the event view, should be button to "Draw Lottery"
- [x] Change the status of the event from "open" to "closed"
- [x] Button should select a random set of winners from the list of entrants
- [x] Update the status of all the entrants

### **Acceptance Criteria (Final)**
- [ ] Send notification to the entrants (related to other user story but can be implemented somewhere here)
- [x] Update database

### PR DESCRIPTION
**Pull request by Kori-Kozicki**
UML, CRC, Storyboard have been updated. Unit tests have been made but haven't done UI tests as I'm still waiting for the merge that fixes that bug. The tests I wrote passed but I haven't been able to actually see the results on the app. So after this PR i'm hoping to connect this to the database and have the results show up on the view entrants list (hoping to do that before the deadline).

### CODE REVIEW

**Review by SuperCreativeUniqueUsername — APPROVED** [review]
looks good to me

File: app/src/main/java/com/example/pygmyhippo/organizer/EventFragment.java [hunk 1]
```diff
@@ -0,0 +1,213 @@
+package com.example.pygmyhippo.organizer;
+
+/*
+This Fragment will display one of the organiser's event they wish to view and give them the option to draw their lottery.
+Purpose is to: Allow the organiser to update their event
+                - Let the organiser see the entrants for that event
+                - Let the organiser draw the lottery for that event
+Contributors: Katherine, Kori
```
Comment (katzhong) [inline app/src/main/java/com/example/pygmyhippo/organizer/EventFragment.java]: NIT: Katharine

File: app/src/main/java/com/example/pygmyhippo/organizer/EventFragment.java [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+
+        closeEventButton.setOnClickListener(new View.OnClickListener(){
+            @Override
+            public void onClick(View view) {
+                // Only do the draw if the event hasn't already been closed
+                if (event.getEventStatus().value.equals("ongoing") && !event.getEntrants().isEmpty()) {
+                    // Change the view of the button
+                    closeEventButton.setBackgroundColor(0xFFA4A8C3);
+                    closeEventButton.setText("Lottery closed");
+                    closeEventButton.setClickable(false);
```
Comment (katzhong) [inline app/src/main/java/com/example/pygmyhippo/organizer/EventFragment.java:154]: just wondering, the organizers can't reopen an event right?

File: app/src/main/java/com/example/pygmyhippo/organizer/EventFragment.java [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+                    closeEventButton.setTextColor(0xFF3A5983);
+                    closeEventButton.setTextSize(20);
+
+                    // Update the status of the event to closed
+                    event.setEventStatus(Event.EventStatus.cancelled);
```
Comment (katzhong) [inline app/src/main/java/com/example/pygmyhippo/organizer/EventFragment.java:159]: just curious, what do the event statuses correspond to again? is cancelled the right status to have here (oragnizer closing event for winners to be drawn)?

File: app/src/main/java/com/example/pygmyhippo/user/EventFragment.java [hunk 4]
```diff
@@ -1,5 +1,16 @@
 package com.example.pygmyhippo.user;
 
+/*
+This Fragment will display one of the events that a User can see after scanning it's QR code
```
Comment (katzhong) [inline app/src/main/java/com/example/pygmyhippo/user/EventFragment.java]: NIT: it's --> its

File: app/src/main/java/com/example/pygmyhippo/user/EventFragment.java [hunk 5]
```diff
@@ -1,5 +1,16 @@
...
+Purposes:
+        - Let the User view the details of the event
+        - Let the User join the event if they wish
+        - If they navigate back to this event, allow them the option to leave the event
+Contributors: Katherine
```
Comment (katzhong) [inline app/src/main/java/com/example/pygmyhippo/user/EventFragment.java]: NIT: Katharine

**Comment by katzhong** [comment]
looks good, left some comments for some minor things :P

File: app/src/main/java/com/example/pygmyhippo/organizer/EventFragment.java (hunk 2 above)
Comment (Kori-Kozicki) [reply app/src/main/java/com/example/pygmyhippo/organizer/EventFragment.java:154]: No. I don't think we need that functionality either. Unless you recall a scenario or story where we do

File: app/src/main/java/com/example/pygmyhippo/organizer/EventFragment.java (hunk 1 above)
Comment (Kori-Kozicki) [reply app/src/main/java/com/example/pygmyhippo/organizer/EventFragment.java:8]: what does NIT stand for?

File: app/src/main/java/com/example/pygmyhippo/organizer/EventFragment.java (hunk 3 above)
Comment (Kori-Kozicki) [reply app/src/main/java/com/example/pygmyhippo/organizer/EventFragment.java:159]: Yah I think the wording we picked is a little misleading. But ongoing means its still opened and cancelled means that its closed. I believe so anyways

**Comment by Kori-Kozicki** [comment]
> looks good, left some comments for some minor things :P

I shall fix those now. After would you care to approve it :)

File: app/src/main/java/com/example/pygmyhippo/organizer/EventFragment.java (hunk 1 above)
Comment (Kori-Kozicki) [reply app/src/main/java/com/example/pygmyhippo/organizer/EventFragment.java]: oop my bad. Spelling is hard

File: app/src/main/java/com/example/pygmyhippo/organizer/EventFragment.java [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+        Random rand = new Random();
+        int winnerNumber = 0;
+        ArrayList<Entrant> entrants = event.getEntrants();
+
+        // If the entrant list is smaller than the winner size, then just set them all as winners
+        if (entrants.size() <= event.getEventWinnersCount()) {
+            // Use the winner number as an index
+            while (winnerNumber < entrants.size()) {
+                // Update all the entrants statuses
+                //TODO: Notifications would probably be sent from here
```
Comment (avreet-14) [inline app/src/main/java/com/example/pygmyhippo/organizer/EventFragment.java:192]: do I need to implement something related to notifications over here too, if the PR gets approved ?

**Comment by avreet-14** [comment]
everything looks good to me

**Review by avreet-14 — APPROVED** [review]

File: app/src/main/java/com/example/pygmyhippo/organizer/EventFragment.java (hunk 6 above)
Comment (Kori-Kozicki) [reply app/src/main/java/com/example/pygmyhippo/organizer/EventFragment.java:192]: That's for the future. But yes when notifications get sent I think there is likely the spot they get sent from. Unless you see a better spot to send them. Either way it's a suggestion that in the future changing status should send notifications.

**Pull request merged**